docs: add third-party storage docs

Vicente-Cheng opened the pull request:
Add third-party storage support docs.

Related: https://github.com/harvester/harvester/issues/1199

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
 ## Use the CSI driver
 
-After successfully configuring these settings, you can utilize the third-party StorageClass. You can apply the third-party StorageClass when creating an empty volume or adding a new block volume to a VM, enhancing your Harvester cluster's storage capabilities.
+After we have installed the CSI driver and configured the Harvester Cluster, we can now use the third-party storage solution in Harvester on serval scenarios.
```
```suggestion
After we have installed the CSI driver and configured the Harvester Cluster, we can now use the third-party storage solution in Harvester on several scenarios.
```

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+Now, we can use the third-party storage solution as the VM image storage.
+
+1. GoTo **Image** and Click **Create**
```
```suggestion
1. Go to **Image** and Click **Create**
```

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+Now, we can use the third-party storage solution as the VM root volume and data volume.
+
+1. GoTo **Virtual Machine** and Click **Create**
+1. GoTo **Volumes** and select the VM image created with the third-party storage solution.
```
```suggestion
1. Go to **Virtual Machine** and click **Create**
1. Go to **Volumes** and select the VM image created with the third-party storage solution.
```

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+Since we support third-party storage solutions, you can create the volume and select the corresponding volume mode.
+
+1. GoTo **Volumes** and Click **Create**
```
```suggestion
1. Go to **Volumes** and click **Create**
```

ihcsim commented on docs/vm/backup-restore.md:
```diff
@@ -26,6 +26,8 @@ VM backups are created from the **Virtual Machines** page. The VM backup volumes
 
 A backup target must be set up. For more information, see [Configure Backup Target](#configure-backup-target). If the backup target has not been set, you’ll be prompted with a message to do so.
 
+The backup only supports the Longhorn v1 Data Engine. If the VM contains any volume from the third-party storage provider, the backup feature will be blocked.
```
```suggestion
The backup only works with the Longhorn v1 Data Engine. If the VM contains any volume from the third-party storage provider, the backup feature will be blocked.
```

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+### Top level architecture
+
+By introducing the Containerized-Data Importer (CDI) into Harvester, Harvester can abstract the VM Image and VM Root Disk path for generic third-party storage solutions. The CDI will help to import the VM Image and VM Root Disk. So, the third-party storage solution can be used as the VM Image and VM Root Disk.
```
```suggestion
By introducing the [Containerized-Data Importer (CDI)](https://kubevirt.io/user-guide/storage/containerized_data_importer/) into Harvester, Harvester can abstract the VM Image and VM Root Disk path for generic third-party storage solutions. The CDI will help to import the VM Image and VM Root Disk. So the third-party storage solution can be used as the VM Image and VM Root Disk.
```

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+#### Storage Profile
+
+The CRD **StorageProfile** is the import concept for the Containerized-Data Importer (CDI). The **StorageProfile** can be used to define individual configuration for different storage providers. The following is the example for the **StorageProfile** for the LVM storage provider. We can use it to know each fields' meaning.
```
> import concept

What does this mean? Or did you mean "important concept"?

ihcsim commented on docs/vm/create-vm.md:
```diff
@@ -152,6 +152,7 @@ A disk can be made accessible via the following types:
 | cd-rom | This type will expose the volume as a cd-rom drive to the VM. It is read-only by default. |
 
 A volume's [StorageClass](../advanced/storageclass.md) can be specified when adding a new empty volume; for other volumes (such as VM images), the `StorageClass` is defined during image creation.
+Since the third-partry storage is supported, you can select the corresponding `StorageClass` and `volumeMode` for the volume. e.g., StorageClass with the NFS storage should use the `Filesystem` volumeMode.
```
```suggestion
If using third-party storage, make sure you choose the correct `StorageClass` and `volumeMode` for the volume. e.g., StorageClass with the NFS storage should use the `Filesystem` volumeMode.
```

ihcsim commented on docs/vm/edit-vm.md:
```diff
@@ -49,6 +49,12 @@ For more details about the network implementation, please refer to the [Networki
 
+:::note
+
+Since the third party storage is supported. You need to ensure the expansion capability of the third-party storage provider before you resize the volume.
```
```suggestion
If using third party storage, you need to ensure that the third-party storage provider supports volume expansion before resizing your volume.
```

jillian-maroket commented on docs/volume/export-volume.md:
```diff
@@ -13,6 +13,12 @@ description: Export volume to image from the Volume page.
 
+:::note
+
+Since the third-party storage is supported, there is the limitation of the export volume. You need to ensure the volume did not attach to any workload (VM/Pods) before exporting the volume to an image or you will see the image is stuck in the `Exporting` status.
```
```suggestion
If the volume is in external storage, ensure that it is not attached to workloads (virtual machines and pods) before starting the export process. This prevents the resulting image from getting stuck in the *Exporting* state.
```

jillian-maroket reviewed (commented):
Initial review done. Please check my comments.

Vicente-Cheng commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,131 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+Currently, the backup feature only supports the Longhorn v1 Data Engine. If the VM contains any volume from the third-party storage provider, the backup feature will be blocked.
+
+#### Cannot convert the attached PVC to the VM Image
```
updated

Vicente-Cheng commented:
>I don't know if we're allowed to link to HEPs. The end user doc should focus on key concepts and operational procedures. Design and implementation details can be discussed only if users are required to understand them. Please check if the topics in this section must really be included.

The topic `Advanced/Internal Topics` I used to introduce the `StorageProfile` concept because this concept is very important once user wants to use their own storage solution. We need to let them know how they should configure.
Because of that, I added this topic.
Or did you think I just mentioned the StorageProfile for something topic like `Advanced`?

WDYT for this section? cc @ihcsim, @bk201, @WebberHuang1118, @jillian-maroket

ihcsim commented:
@Vicente-Cheng I am not sure if we are responsible for providing storage profile examples in our documentation. IIUC, we just need users to know it will work out-of-box because Harvester integrated with CDI, right? It's not mandatory to use storage profiles in Harvester. For some well-known provisioners, CDI might even auto-create the storage profiles based on their storage classes.

If this is just for the user's FYI, would something like this make sense:

"By integrating Harvester with CDI, you can now utilize the CDI API to create custom storage profiles to simplify your data volumes definition. Storage profiles allow multiple data volumes to share the same provisioner settings.

Here is an example of a LVM storage profile:

# ....

For more information on storage profiles, refer to the [CDI documentation](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md)."

Other suggestions on the same section:

1. Omit the "Top level architecture" section. HEP sounds like a better home for it
2. Omit the clone strategies. We are not responsible for explaining CDI concepts

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,120 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+More detailed can be found in the HEP [Third Party Storage Support](https://github.com/harvester/harvester/blob/master/enhancements/20250203-third-party-storage-support.md).
+
+### Notable Configuration Options
```
This heading is not necessary. Please remove it.

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,120 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+### Notable Configuration Options
+
+#### Storage Profile
+
+The CRD **StorageProfile** is the important concept for the Containerized-Data Importer (CDI). The **StorageProfile** can be used to define individual configuration for different storage providers. The following is the example for the **StorageProfile** for the LVM storage provider. We can use it to know each fields' meaning.
```
```suggestion
### Storage Profiles

You can now use the CDI API to create custom [storage profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) that simplify definition of data volumes. Storage profiles allow multiple data volumes to share the same provisioner settings.

The following is an example of an LVM storage profile:
```

jillian-maroket replied:
I made a few changes to Ivan's suggested wording and added the link to the CDI doc.

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +57,120 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  storageClass: lvm-node-1-striped
+```
+
+- claimPropertySets: The claimPropertySets field is used to define the accessModes and volumeMode for the StorageProfile.
+- snapshotClass: The snapshotClass field is used to define the VolumeSnapshotClass for the StorageProfile.
+- cloneStrategy: The cloneStrategy field is used to define the clone strategy for the StorageProfile.
+
+The cloneStrategy is associated with the [efficient clone](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/efficient-cloning.md).
+
+Efficient clone could be classified into three categories:
+- *CSI volume cloning* - Leverage the clone capability of the storage provider. The corresponding cloneStrategy is **csi-clone**.
+- *Smart Cloning* (snapshot-based cloning) - Leverage the snaphsot capability of the storage provider. The corresponding cloneStrategy is **snapshot**.
+- *host-assisted cloning* - Depends on directly data copy from source pod (volume) to target pod (volume). The corresponding cloneStrategy is **copy**.
````
```suggestion
For more information, see [Storage Profiles](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md) in the CDI documentation.
```

jillian-maroket replied:
The draft field descriptions do not provide a lot of information. Also, @ihcsim [suggested](https://github.com/harvester/docs/pull/728#issuecomment-2749602168) removing the clone strategies part.

jillian-maroket commented on docs/vm/create-vm.md:
```diff
@@ -152,6 +152,7 @@ A disk can be made accessible via the following types:
 | cd-rom | This type will expose the volume as a cd-rom drive to the VM. It is read-only by default. |
 
 A volume's [StorageClass](../advanced/storageclass.md) can be specified when adding a new empty volume; for other volumes (such as VM images), the `StorageClass` is defined during image creation.
+If using third-party storage, make sure you choose the correct `StorageClass` and `volumeMode` for the volume. e.g., StorageClass with the NFS storage should use the `Filesystem` volumeMode.
```
```suggestion
If you are using external storage, ensure that the correct **StorageClass** and **Volume Mode** are selected. For example, a volume with the **nfs-csi** StorageClass must use the **Filesystem** volume mode.
```

jillian-maroket commented on docs/vm/edit-vm.md:
```diff
@@ -49,6 +49,12 @@ For more details about the network implementation, please refer to the [Networki
 
+:::note
+
+If using third party storage, you need to ensure that the third-party storage provider supports volume expansion before resizing your volume.
```
```suggestion
If you are using external storage, ensure that the storage provider supports volume expansion before you resize volumes.
```

jillian-maroket reviewed (commented):
I reviewed the parts that were skipped in the initial review. Note that I agree with Ivan's suggestions, which are reflected in my recommended wording and structural changes.

innobead commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,120 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 
 
-![rook-ceph-volume-external](/img/v1.2/advanced/rook-ceph-volume-external.png)
+- A known issue in the CDI prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.
```
Suggest adding the upstream issue for reference.

Vicente-Cheng replied:
Could we use limitation instead of issue?
I will create the upstream issue, but that might need some discussion.

innobead replied:
Sounds good.

Vicente-Cheng commented:
> I reviewed the parts that were skipped in the initial review. Note that I agree with Ivan's suggestions, which are reflected in my recommended wording and structural changes.

All updated!
Thanks for the review, please recheck.

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -8,23 +8,29 @@ title: "Third-Party Storage Support"
-:::note
+:::important
+
+The matrix below shows the supported capabilities of the validated CSI drivers.
```
Does it make sense to include a href here to the SUSE certified storage for virtualization? E.g., _"For more information on SUSE certified 3rd-party storage, see https://www.suse.com/product-certification/suse-certified/virtualization-certification/."_

Vicente-Cheng replied:
Thanks @ihcsim,
Actually, I did not know this page. Do we already have the certified storage?

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +57,105 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+### Limitations
 
-With these configurations in place, your Harvester cluster is ready to make the most of the third-party storage integration.
+- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 
```
I think we should link this to the "VM Backup Compatibility" added in https://github.com/harvester/docs/pull/731, to provide user with more context. E.g., _"For more information, see [VM Backup Compatibility](#vm-backup-compatibility)."_

ihcsim reviewed (approved):
LGTM - just 2 minor suggestions.

WebberHuang1118 reviewed (approved):
LGTM, thanks.

votdev commented on docs/advanced/csidriver.md:
```diff
@@ -8,23 +8,31 @@ title: "Third-Party Storage Support"
+| NFS | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10004; | &#10006; | &#10006; |
 
-The Virtual Machine (VM) image provisioner in Harvester still relies on Longhorn. Before version 1.2.0, Harvester exclusively supported Longhorn for storing VM data and did not offer support for external storage as a destination for VM data.
+For more information on SUSE certified 3rd-party storage. See [SUSE Certified Storage for Virtualization](https://www.suse.com/product-certification/suse-certified/virtualization-certification/).
```
This sentence is weird and looks like to be incomplete.

votdev commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,107 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+Backup currently only works with the Longhorn v1 Data Engine. If you are using other storage providers, you can skip the **Backup VolumeSnapshot Class Name** configuration.
+
+For more information, see [VM Backup Compatibility](../../versioned_docs/version-v1.4/advanced/csidriver.md#vm-backup-compatibility).
```
I think this link is wrong. It should point to the latest version.
```suggestion
For more information, see [VM Backup Compatibility](./csidriver.md#vm-backup-compatibility).
```

votdev commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,107 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+- Backup support is currently limited to Longhorn V1 Data Engine volumes. Harvester is unable to create backups of volumes in external storage. 
 
-![rook-ceph-volume-external](/img/v1.2/advanced/rook-ceph-volume-external.png)
+- There is a limitation in the CDI prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.
```
```suggestion
- There is a limitation in the CDI which prevents Harvester from converting attached PVCs to virtual machine images. Before exporting a volume in external storage, ensure that the PVC is not attached to workloads. This prevents the resulting image from getting stuck in the *Exporting* state.
```

Vicente-Cheng commented:
Hi @jillian-maroket, 

I have added the appendix on how to deploy the NFS-related sample.
Could you help recheck this part? Thanks!

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+:::
+
+- Install with csi-driver-nfs helm chart
```
```suggestion
1. Install the driver using the `csi-driver-nfs` Helm chart.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  $ helm install csi-driver-nfs csi-driver-nfs/csi-driver-nfs --namespace kube-system --version v4.10.0
+  ```
+
+- create the StorageClass for NFS
````
```suggestion
1. Create the StorageClass for NFS.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+    - nfsvers=4.2
+  ```
+
+After that, you can create VM Image/Root Volume/Data Volume with the StorageClass `nfs-csi`.
````
```suggestion
  Once created, you can use the StorageClass to create virtual machine images, root volumes, and data volumes.
```

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+After that, you can create VM Image/Root Volume/Data Volume with the StorageClass `nfs-csi`.
+
+### How to deploy the AllInOne NFS Service + CSI driver
```
```suggestion
### Deploy the Combined NFS Service and CSI Driver
```

jillian-maroket replied:
There are no search results for "AllInOne NFS Service" so I assume this is not an official term.

Vicente-Cheng replied:
Yeah, you were correct!
Updated!

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+### How to deploy the AllInOne NFS Service + CSI driver
+
+- clone the AllInOne repo https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner
```
```suggestion
1. Clone the [nfs-ganesha-server-and-external-provisioner repository](https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner).
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  $ git clone https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner 
+  ```
+
+- change the directory to `nfs-ganesha-server-and-external-provisioner/deploy/kubernetes` and modify the deployment.yaml file as follows:
````
```suggestion
1. Change the directory to `nfs-ganesha-server-and-external-provisioner/deploy/kubernetes` and modify the `deployment.yaml` file as follows:
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+              path: <your-nfs-export-path>
+  ```
+
+- apply the deployment.yaml file
````
```suggestion
1. Apply the `deployment.yaml` file.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  ```
 
-## Use the CSI driver
+- apply the RBAC
````
```suggestion
1. Apply the RBAC configuration.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  ```
 
-After successfully configuring these settings, you can utilize the third-party StorageClass. You can apply the third-party StorageClass when creating an empty volume or adding a new block volume to a VM, enhancing your Harvester cluster's storage capabilities.
+- create the StorageClass for NFS
````
```suggestion
1. Create the StorageClass for NFS.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  ```
 
-With these configurations in place, your Harvester cluster is ready to make the most of the third-party storage integration.
+- The above StorageClass named example-nfs, so we need to update the corresponding StorageProfile CRD as below
````
```suggestion
1. Update the StorageProfile CRD that corresponds to the StorageClass you created.

  In the following example, the StorageClass is `example-nfs`.
```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+  ```
 
-![rook-ceph-volume-external](/img/v1.2/advanced/rook-ceph-volume-external.png)
+After that, you can create VM Image/Root Volume/Data Volume with the StorageClass `example-nfs`.
````
```suggestion
  Once created, you can use the StorageClass to create virtual machine images, root volumes, and data volumes.
```

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -8,23 +8,31 @@ title: "Third-Party Storage Support"
-- Support snapshot
-- Support clone
-- Support block device
+For the Harvester functions to work well, the third-party CSI driver should support the following capabilities:
+- Volume expansion, for online resizing of the volume
+- Volume snapshot, for taking snapshots of the volume/VM
+- Volume clone, for cloning the volume/VM
 - Support Read-Write-Many (RWX), for [Live Migration](../vm/live-migration.md)
```
```suggestion
To enable Harvester to function well, use CSI drivers that support the following capabilities:
- Volume expansion (online resizing)
- Snapshot creation (volume and virtual machine snapshots)
- Cloning (volume and virtual machine clones)
- Usage of Read-Write-Many (RWX) volumes for [Live Migration](../vm/live-migration.md)
```

jillian-maroket replied:
This suggestion was missed.

jillian-maroket reviewed (commented):
Review done

Vicente-Cheng commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+![convert-pvc-to-image-stuck](/img/v1.5/advanced/convert-pvc-to-image-stuck.png)
+
+## Appendix
```
Sounds good! Thanks for the explanation!

ihcsim commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +59,200 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+![convert-pvc-to-image-stuck](/img/v1.5/advanced/convert-pvc-to-image-stuck.png)
+
+### How to deploy the NFS CSI driver
```
Why do we need to include this in our docs? It feels a bit out of place. If any of the upstream csi nfs instructions changed, our documentation will become stale. Also, the provisioner at https://github.com/kubernetes-sigs/nfs-ganesha-server-and-external-provisioner seems very old. If we are going to recommend it in our docs, we need to make sure their images are safe and updated.

Alternately, what do you of converting this content into a blog post or recording or kubecon talk proposal where we showcase how to use nfs csi driver and storage profile in 1.5? I can help with this (especially, kubecon proposal :wink:), if needed.

ihcsim replied:
If we really want to include this in our docs, maybe consider just add a href to https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/deploy/example/README.md and call it done?

ihcsim replied:
yeah, i prefer not to include it in our docs. many of its closed issues indicated that the project is no longer maintained. any reasons why we don't want to link to the csi nfs official docs at https://github.com/kubernetes-csi/csi-driver-nfs/blob/master/deploy/example/README.md?

Vicente-Cheng replied:
OK, let's remove!

Vicente-Cheng commented:
Hi folks,
I added the two parts. Please help review them again.

1. Added the validated CSI driver. (Even the same storage solution would use a different CSI driver, so I specify the CSI driver we validated)
2. I added the NFS server-side configuration (squash related) to give the user a hint on how to configure it to make things go well.

Vicente-Cheng commented:
Hi folks,
I also added the known issue part for [[DOC] Adjusting CDI filesystemOverhead to prevent scratch volume space shortage](https://github.com/harvester/harvester/issues/8039#top)

Please help review them. Thanks!

WebberHuang1118 reviewed (approved):
LGTM, thanks.

jillian-maroket commented on docs/advanced/csidriver.md:
```diff
@@ -51,25 +65,182 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+
+## Known Issues
+
+### 1. infinite image download loop
+
+Reference issue: https://github.com/harvester/harvester/issues/7993 
+
+This issue is related to the scratch volume, which is temporarily used to store the image data and is created by CDI. We might see the following error logs on the importer-prime-xxx pod.
```
```suggestion
### 1. Infinite Image Download Loop

The image download process loops endlessly when the StorageClass for the image uses the LVM CSI driver. This issue is related to the scratch volume, which is created by CDI and is used to temporarily store the image data. When the issue exists in your environment, you might find the following error messages in `importer-prime-xxx` pod logs:

```

jillian-maroket replied:
We usually mention the related issue after the issue description.

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +65,182 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+runtime.main
+```
+
+The above log shows that the filesystem created with scratch volume is not enough to store the image data. That's why we see the error `no space left on device`. The scratch volume is created to be the same size as the target volume but adds some filesystem overhead. The filesystem overhead default is 0.055, which means 5.5%. It usually works in most cases. However, if the image size is quite sma ...
````
```suggestion
The message `no space left on device` indicates that the filesystem created using the scratch volume is not enough to store the image data. CDI creates the scratch volume based on the size of the target volume, but some space is lost to filesystem overhead. The default overhead value is `0.055` (equivalent to 5.5%), which is sufficient in most cases. However, if the image size is less than 1 GB and its virtual size is very close to the image size, the default overhead is likely to be insufficient.

The workaround is to increase the filesystem overhead to 20% using the following command:

```

jillian-maroket commented on docs/advanced/csidriver.md:
````diff
@@ -51,25 +65,182 @@ With the kubeconfig of the Harvester cluster, you can install the third-party CS
+# kubectl patch cdi cdi --type=merge -p '{"spec":{"config":{"filesystemOverhead":{"global":"0.2"}}}}'
+```
+
+The above command will set the filesystem overhead to 20%. After that, we should see the image downloading is complete. And don't worry about the higher filesystem overhead. It will not affect the image PVC size. This PVC (scratch volume) will be deleted after the image is imported.
+
+For more detailed reasons, you can see this [comment](https://github.com/harvester/harvester/issues/7993#issuecomment-2790260841).
````
```suggestion
The image should be downloaded once the filesystem overhead is increased.

:::note

Increasing the overhead value does not affect the image PVC size. The scratch volume is deleted after the image is imported.

:::

Related issue: [#7993](https://github.com/harvester/harvester/issues/7993) (See this [comment](https://github.com/harvester/harvester/issues/7993#issuecomment-2790260841).)
```

jillian-maroket reviewed (commented):
I reviewed the new text.

Vicente-Cheng commented on docs/advanced/csidriver.md:
```diff
@@ -8,24 +8,38 @@ title: "Third-Party Storage Support"
+- LVM: lvm.driver.harvesterhci.io
+- NFS: nfs.csi.k8s.io
+
+For more information on SUSE certified 3rd-party storage, see [SUSE Certified Storage for Virtualization](https://www.suse.com/product-certification/suse-certified/virtualization-certification/).
```
yeah, I thought we could remove this link at this moment.

WebberHuang1118 reviewed (approved):
LGTM, thanks.
